### PR TITLE
Issue/974 restore post action

### LINKED ISSUE #974: Restore post action
We need a `RESTORE_POST` action to restore posts that are currently trashed. Once a post is restored its value in DB needs to be updated.

This action needs to be implemented for both REST and XML-RPC sites. Calypso seems to be using `v1.1/sites/$siteId/posts/$postId/restore` and as far as I can tell there is no specific endpoint for XML-RPC sites, but I suspect it is just a status change, so we should be able to re-dispatch the `PUSH_POST` action for XML-RPC sites after updating the status of the post.

It'd be good if we have a separate `CauseOfChange` for restore action, but if the XML-RPC uses `PUSH_POST`, I can't see a big issue with using `CauseOfChange.UpdatePost`. In this case, the REST API could also possibly use `PUSHED_POST` for the return event to make it easier and also keep it consistent between two network layers.

P.S: This issue is written in a way to keep things consistent with how PostStore works and admittedly not the best way to do this.

### PR DESCRIPTION
**Pull request by khaykov**
Fixes #974 

This PR adds a `RESTORE_POST` action that allows us to restore trashed remote posts. I this as an opportunity to learn about XmlRpc and recent changes to Post list. I hope understood everything correctly, but please let me know if I missed something. I added comments to a couple of places I'm not 100% sure about.

Notes:
I went with the separate `CauseOfChange` for post-restoration, which should make life a bit easier on a client side.

XmlRpc client is using `EDIT_POST` endpoint, and then fetches updated post from the server before emitting a successful event.  The process is really similar to on in post updater flow, so there is a bit of code duplication.

There is catch with statuses of restored posts on the self-hosted websites - we have do not know the status of a post before it was trashed, so all the restore post will become published. I'm not a fan of this approach, and wonder, maybe we can set the status of restored posts to `private` and notify user about this in a client?

### CODE REVIEW

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java [hunk 1]
```diff
@@ -717,6 +737,25 @@ private void handlePushPostCompleted(RemotePostPayload payload) {
...
+            event.error = payload.error;
+            emitChange(event);
+        } else {
+            if (payload.site.isUsingWpComRestApi() || syncingNonWpComPost) {
+                restorePost(payload.post);
+            } else {
+                // XML-RPC responds to post restore request with status boolean
+                // Update the post locally to reflect its published state, and request a fresh copy
+                // from the server to ensure local copy matches server
+                PostSqlUtils.insertOrUpdatePostOverwritingLocalChanges(payload.post);
```
Comment (khaykov) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java]: `handlePushPostCompleted` adds a post to the DB before fetching a 'true' copy from the server and delivering it to subscribers. I used the same approach here, but I'm not 100% sure why we add it to the DB twice when only notifying observers once.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java [hunk 2]
```diff
@@ -759,6 +798,17 @@ private void removeAllPosts() {
...
     }
 
+    private void restorePost(PostModel postModel) {
+        int rowsAffected = PostSqlUtils.insertOrUpdatePostOverwritingLocalChanges(postModel);
+        CauseOfOnPostChanged causeOfChange =
+                new CauseOfOnPostChanged.RestorePost(postModel.getId(), postModel.getRemotePostId());
+        OnPostChanged onPostChanged = new OnPostChanged(causeOfChange, rowsAffected);
+        emitChange(onPostChanged);
+
+        mDispatcher.dispatch(ListActionBuilder.newListItemsChangedAction(
```
Comment (khaykov) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java]: I was looking really hard for subscribers to this in a client but was not able to find anything. Am I missing something?

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java (hunk 1 above)
Comment (aforcier) [reply fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java]: Yep, that's about right. We inherited this approach from WPAndroid during the FluxC migration (the part where no callback was triggered until the second network call, the fetch, had completed) and kept it the same (it made some sense since the calls would behave as REST does).

I don't have all the context, but I'm not sure an extra callback would be the best approach. @oguzkocer could you elaborate on what this means?

> it ends up mostly hiding potentially important information from the client

Assuming you're referring to the error handling becoming messy, it seems to me that having an extra callback that's only triggered for XML-RPC sites (which we generally try to abstract away as much as possible) just seems like trading one complexity for another. It might be better instead to improve FluxC's internal handling here, to make sure the end result is reported properly ('push failed' or 'fetch failed'), but continue to abstract away XML-RPC's unfortunate two-step post push process.

> It's really hard to find a good approach with XML-RPC API 🤷‍♂️

Basically comes down to this 🙂

Comment (oguzkocer) [reply fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java]: > I don't have all the context, but I'm not sure an extra callback would be the best approach. @oguzkocer could you elaborate on what this means?

@aforcier What I was trying to say was, I'd have preferred if we had 2 `OnPostChanged` events after a `PUSH_POST` action for XML-RPC sites, once when we complete the push and update the DB with the pushed post details and once after we fetch the recently pushed post. That way, the error handling becomes clear, because we know whether the push has failed or the fetch after has, because if it's the fetch, in most cases it should be safe to ignore it. We don't really care how many events are emitted due to an action due to FluxC's design, so I don't think there is a significant trade-off for this approach.

_All that ^ is of course assuming that getting 200 from the push request does mean that the post we just pushed will most likely matches the server version, because otherwise, we'll update the UI prematurely._

Having said that, I wasn't actually suggesting a change here, either way we are dealing with edge cases, so there is no reason to go ahead with a change for something that has been working without any major issues so far. I just like to be through with my thought whenever I am reviewing a PR as I have found that to work better for me.

> Assuming you're referring to the error handling becoming messy, it seems to me that having an extra callback that's only triggered for XML-RPC sites (which we generally try to abstract away as much as possible) just seems like trading one complexity for another.

100% agreed. I thought I said something similar in my comment, but reading it back, I guess not 🙈 

> It might be better instead to improve FluxC's internal handling here, to make sure the end result is reported properly ('push failed' or 'fetch failed'), but continue to abstract away XML-RPC's unfortunate two-step post push process.

That is certainly another good approach. The trade-off for that one would be that the client needs to know about an error may not mean an error and needs to be handled accordingly which is not a library type thing to do, but with WordPress, even after all of our best attempts at it, it's almost impossible to do anyway. :)

@khaykov @aforcier My suggestion for this PR would be to follow this PR as is and if we want to go with one approach or the other, we should open an issue for it and open a separate PR that updates both the push action as well as the restore one. What do you think?

**Comment by khaykov** [comment]
@oguzkocer Thanks for review and some insight into flux-c :) 
I did some refactoring and added connected tests. Ready for another round!

**Comment by khaykov** [comment]
@oguzkocer Thanks for taking another look! I think I addressed all the comments :)
